src/Control/Monad/Infer.hs

```diff
@@ -162,12 +163,20 @@ instantiate tp@(TyPi (Implicit v _) ty) = do
     TgName n _ -> TgName n <$> gen
   let map = Map.singleton v var
 
-      squish f (Just x) = Just (x . f)
-      squish f _ = Just f
-
       appThisTy e = ExprWrapper (TypeApp var) e (annotation e, apply map ty)
   (k, _, t) <- instantiate (apply map ty)
   pure (squish appThisTy k, tp, t)
+instantiate tp@(TyPi (Anon co) od@dm) = do
+  (wrap, _, dm) <- instantiate dm
+  let cont = fromMaybe id wrap
+  var <- fresh
+  let ty = TyPi (Anon co) dm
+      lam :: Expr Typed -> Expr Typed
+      lam e
+        | ann <- annotation e
+        = Fun (PType (Capture (TvName var) (ann, co)) co (ann, co)) (cont (App e (VarRef (TvName var) (ann, co)) (ann, od))) (ann, ty)
```

Comment:
I assume this is the only way, but it seems to create a lot of noise: `2 + 3` expands to `(fun x -> fun y -> (+) x y) 2 3` for instance, which is a little un-ideal.